### ISSUE TITLE
Reload / Restart all Kopf handlers when a configmap changes.

### ISSUE THREAD

**Issue by smileisak** [nolar/kopf#515]
## Question

What is the best approach to restart/reload the operator when a config map has changed?

The idea behind this is when a configuration inside a configmap changes, Kopf detects the event and reload itself.
## Checklist

- [x] I have read the [documentation](https://kopf.readthedocs.io/en/latest/) and searched there for the problem
- [x] I have searched in the [GitHub Issues](https://github.com/nolar/kopf/issues?utf8=%E2%9C%93&q=) for similar questions

## Keywords

`reload`, `restart`, `configmaps`, `configurations`

**Comment by eshepelyuk** [comment]
Hello, 

1. The most obvious solution - is to restart your operator pods, so it will reconfigure itself on restart.

       $ kubectl rollout restart deployment $YOUR_DEPLOYMENT_NAME

1. Better option - create a kopf timer that periodically checks if file mounted from config map had changed ( I assumed your operator uses some config file from a ConfigMap). As soon as ConfigMap changes - the content will be available inside oeprator's pod, so your timer could check either content hash or file attributes to understand if your operator need to reconfigure itself. Then, timer could store new config, so handlers will use new config the next time they will be invoked.

1. Also on ConfigMap change - you could exit your operator with `sys.exit(1)`. K8S should recreate containers in this case, that will virtually result in option #1, but rather automated.

**Comment by smileisak** [comment]
@eshepelyuk thank you for those ideas.
 
Both solution 2 and 3 need a timer to be performed, I have operators on production using `v0.26` and I cannot upgrade the version of Kopf to `v0.27` to have timers functionally due to this issue #375 

Besides, timers are not event-driven handlers. So when the configuration changes, we need to wait until the time comes to detect those changes.

**Comment by eshepelyuk** [comment]
@nolar @smileisak I suppose this project should be used for such cases, no additional development in Kopf needed.

https://github.com/stakater/Reloader

**Comment by smileisak** [comment]
@eshepelyuk, I've used Reloader for my case and It is doing the job like a charm! Thank you again.

**Comment by MartinEmrich** [comment]
Instead of opening a new issue: I try to achieve this without actually restarting the controller pod. Is this possible?

Heres some "do-what-i-mean"-pseudocode, the options unknown to me in Caps:
````
@ kopf.on.event("configmap", WATCH_ONLY_THIS="the-name-of-my-configmap", EVENT_ONLY="changed")
def configmap_event_watcher(logger, **kwargs):
    logger.info("Configmap XXX changed, reconciling all resources...")
    kopf.TRIGGER_RESUME_ON_ALL_RESOURCES(xxx)
````
Doing the rollout-on-change thing outside of the controller might be effective, but just feels a bit brute-force to me.

Thanks!

**Comment by nolar** [comment]
I'm not sure what is meant here by "changed" — Kubernetes does not send this kind of information. But you can replace the on-event with on-update, for example — Kopf keeps track of changes (not Kubernetes).

The "triggering of resume" is impossible: you have no content of those resources you want to resume. Normally, handlers get the values from the Kubernetes events (or initial listing), and those values are discarded once the handlers are finished.

But you can use indexes to keep track of all involved objects and their content (here, keeping the full body; to save memory, reduce to something meaningful only). And every time the configmap changes, you can start some activities on all (or some) remembered objects. 

In this example, the activities are executed in parallel in the background, so that you can leave the on-event instantly.

So, it might look brute-force'ish, but probably it is the only way to achieve that nicely.

```python
import kopf
import time
import threading


# Mind the RBAC permissions.
@kopf.index('involvedresource1')
@kopf.index('involvedresource2')
@kopf.index(category='resumable')
@kopf.index(kopf.EVERYTHING, when=lambda name, **_: 'xyz' in name)
def involved_resources(name, namespace, resource, body, **_):
    return {(resource, namespace, name): body}


def configmap_filter(name, **_):
    return name == 'the-name-of-my-configmap'

# note: the indexed record will always have 1 value, never 0, never 2+ (because the key is unique).
# this is why it is safe to unfold it into `(body,)`.
# {(r1, ns1, nam1): [body1],
#  (r2, ns2, nam2): [body2],
#  ...}
@kopf.on.event("configmap", when=configmap_filter)
async def configmap_event_watcher(logger, spec, involved_resources: kopf.Index, **kwargs):
    logger.info("Configmap XXX changed, reconciling all resources...")
    print(involved_resources)  # for preview

    for (obj_res, obj_ns, obj_name), (obj_body,) in involved_resources.items():
        asyncio.create_task(reconcile_async(obj_res, obj_ns, obj_name, obj_body))
        threading.Thread(target=reconcile_sync, args=(obj_res, obj_ns, obj_name, obj_body)).start()

async def reconcile_async(resource, namespace, name, body):
    await asyncio.sleep(10)  # do something (but async!)

def reconcile_sync(resource, namespace, name, body):
    time.sleep(10)  # do something (even sync!)
```

I didn't test it. Just expressing the idea. 

Mind the resource control: if the configmap changes too often, it will spawn many tasks that will be conflicting over each other. You probably want to keep track of all running tasks in the configmap's `memo`.

---

Another way to keep track of all objects is to run daemons for each of them. This is essentially the same as the solution above, but the task orchestration is done by Kopf. You only need to signal the daemons when they need to wake up:

```python
import kopf
import threading

@kopf.on.startup()
def start(memo, **_):
    memo.cnd = threading.Condition()

def configmap_filter(name, **_):
    return name == 'the-name-of-my-configmap'

@kopf.on.event("configmap", when=configmap_filter)
def configmap_event_watcher(logger, spec, memo, **kwargs):
    logger.info("Configmap XXX changed, summoning all daemons...")
    with memo.cnd:
        memo.cnd.notify_all()


# Mind the RBAC permissions.
@kopf.daemon('involvedresource1')
@kopf.daemon('involvedresource2')
@kopf.daemon(category='resumable')
@kopf.daemon(kopf.EVERYTHING, when=lambda name, **_: 'xyz' in name)
def sleeper(stopped, memo, name, namespace, resource, body, **_):
    while not stopped:
        with memo.cnd:
            awakened = memo.cnd.wait(timeout=30)
        if awakened and not stopped:
            print(f'I am {namespace}/{name}')
            time.sleep(10)  # do something
```

Again, I did not test it. Just the idea to look into.

An async daemon would be better:

* First, it could be cancelled at any time. The sync daemon here can wait up to 30 seconds on resource deletion/operator exiting.
* Second, asyncio tasks are lightweight, much more lightweight than threads. If you have hundreds or thousands of resource objects to serve, consider making the daemons async.

The async daemon would be like this:

```python
import kopf
import asyncio

@kopf.on.startup()
def start(memo, **_):
    memo.cnd = asyncio.Condition()

def configmap_filter(name, **_):
    return name == 'the-name-of-my-configmap'

@kopf.on.event("configmap", when=configmap_filter)
async def configmap_event_watcher(logger, spec, memo, **kwargs):
    logger.info("Configmap XXX changed, summoning all daemons...")
    async with memo.cnd:
        memo.cnd.notify_all()


# Mind the RBAC permissions.
@kopf.daemon('involvedresource1')
@kopf.daemon('involvedresource2')
@kopf.daemon(category='resumable')
@kopf.daemon(kopf.EVERYTHING, when=lambda name, **_: 'xyz' in name)
async def sleeper(stopped, memo, name, namespace, resource, body, **_):
    while True:
        async with memo.cnd:
            await memo.cnd.wait()  # <- infinite sleep, but cancellable
        print(f'I am {namespace}/{name}')
        await asyncio.sleep(10)  # do something async
```

There are trade-offs between the 1st and the 2nd solution: the daemons require finalizers, so Kopf will block all involved resources from deletion. If this is not desired, then the 1st way is better.

**Comment by MartinEmrich** [comment]
@nolar Wow thanks for that in-depth explaination. I did not know about *kopf.Index* or the *when=* parameter, that did the trick.

As all I do is a change to a deployment, I omitted all that async/await code and kept it short, this works for me now.

````
def configmap_filter(name, **_):
    return name == 'image-versions'

@kopf.index('mycluster')
def all_clusters(name, namespace, resource, body, patch, logger, **_):
    return {(resource, namespace, name): (body, patch)}

def reconcile_all(involved_resources, logger) -> None:
    for (resource, namespace, name), ((body, patch), ) in involved_resources.items():
        handle_mycluster(body=body, namespace=namespace, name=name, logger=logger, patch=patch)

@ kopf.on.event("configmap", when=configmap_filter)
def configmap_event_watcher(logger, all_myclusters: kopf.Index, type, **kwargs):
    if (type == 'MODIFIED'):
        logger.info("Versions ConfigMap changed, triggering reconcile...")
        update_image_versions()
        reconcile_all(all_myclusters, logger=logger)
````

For future readers: Note that in update_image_versions, I pull the configmap contents via Kubernetes API, I assume that mounting it as a volume into the controller pod would just be too slow here.
Actually, I'll even change it later to use the body delivered in the event watcher.....

`handle_mycluster` is already my resume/create handler, it gets reused here.

**Comment by nolar** [comment]
@MartinEmrich I would like to add a couple notes:

1. Remembering the `patch` object is not very useful. Kopf will not enforce that patch and will not apply it after the handler (indexer) is done. So, if it is modified in some background processes, it will be discarded. Instead, you should better patch with your own call to the API (via pykube-ng or any other preferred client library — using namespace/name as a reference).

2. The filter for MODIFIED can be moved to the callback too — it can save a few lines (and produce less noise in the logs):

```python
def configmap_filter(name, type, **_):
    return name == 'image-versions' and type == 'MODIFIED'
```

But keep in mind that it is not reliable. Kubernetes has its own understanding of what MODIFIED means. This might differ from your intuitive expectations. E.g.:

Sometimes, Kubernetes drops the connection and Kopf reconnects and continues streaming; but in some cases (e.g. "410 Gone" on the resource version), Kopf will re-list the whole resource kind (`type is None`) — and those resources can contain updates.

More often, Kubernetes sends the MODIFIED events for every little unrelated change of any piece of the resource, such as its status, metadata, annotations, labels, whatnot. Including those imposed by Kopf-based and maybe some other 3rd-party controllers. This might trigger too many reconciliations with no actual reason for that.

**Comment by MartinEmrich** [comment]
@nolar  Thanks again for your hints and your concerns! 

I removed the patch field, indeed it was not really used here. I also incorporated the "type" filter, still works fine.

For my use case (the configmap only contains two container image tags for the deployments managed by my controller), perfect reliability or noise is not an issue: if there are "bogus" modification events, the deployment does not change, so nothing happens.
